utils/generate-test-checks.py

```diff
@@ -0,0 +1,402 @@
+#!/usr/bin/env python3
+"""A script to generate FileCheck statements for mlir unit tests.
+
+This script is a utility to add FileCheck patterns to an mlir file.
+
+NOTE: The input .mlir is expected to be the output from the parser, not a
+stripped down variant.
+
+Example usage:
+$ generate-test-checks.py foo.mlir
+$ mlir-opt foo.mlir -transformation | generate-test-checks.py
+$ mlir-opt foo.mlir -transformation | generate-test-checks.py --source foo.mlir
+$ mlir-opt foo.mlir -transformation | generate-test-checks.py --source foo.mlir -i
+$ mlir-opt foo.mlir -transformation | generate-test-checks.py --source foo.mlir -i --source_delim_regex='gpu.func @'
+
+The script will heuristically generate CHECK/CHECK-LABEL commands for each line
+within the file. By default this script will also try to insert string
+substitution blocks for all SSA value names. If --source file is specified, the
+script will attempt to insert the generated CHECKs to the source file by looking
+for line positions matched by --source_delim_regex.
+
+The script is designed to make adding checks to a test case fast, it is *not*
+designed to be authoritative about what constitutes a good test!
+"""
+
+# Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
```

Comment:
I'm adding a copy of this script because I had to make a few tweaks for it to work properly with out IR. The upstream version doesn't like `#ttg.shared_memory` as it thinks `ttg.shared_memory` should be a symbol name defined in the top level.